cmd/internal/database/valkey/valkey.go

```diff
@@ -0,0 +1,362 @@
+package valkey
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/metal-stack/backup-restore-sidecar/cmd/internal/database/leaderelection"
+	"github.com/metal-stack/backup-restore-sidecar/cmd/internal/utils"
+	"github.com/metal-stack/backup-restore-sidecar/pkg/constants"
+	"github.com/redis/go-redis/v9"
+	"github.com/spf13/afero"
+)
+
+const (
+	valkeyDumpFile = "dump.rdb"
+)
+
+type Valkey struct {
+	log      *slog.Logger
+	executor *utils.CmdExecutor
+	datadir  string
+
+	client *redis.Client
+
+	clusterMode     bool
+	clusterSize     int
+	statefulSetName string
+	password        string
+
+	leaderElection       *leaderelection.LeaderElection
+	leaderElectionCancel context.CancelFunc
+}
+
+func (db *Valkey) Check(context.Context) (bool, error) {
+	empty, err := utils.IsEmpty(db.datadir)
+	if err != nil {
+		return false, err
+	}
+	if empty {
+		db.log.Info("data directory is empty")
+	}
+	return empty, nil
+}
+
+func (db *Valkey) Recover(ctx context.Context) error {
+	dump := path.Join(constants.RestoreDir, valkeyDumpFile)
+	if _, err := os.Stat(dump); os.IsNotExist(err) {
+		return fmt.Errorf("restore file not present: %s", dump)
+	}
+
+	if db.clusterMode {
```

Comment:
Probably it's better to reduce code intention by negating this expression, the `for` loop in this function has explicit returns anyway so last line of this function does not need to be repeated.


```suggestion
	if !db.clusterMode {
```